File: backend/src/main/java/com/hgw/gestionale/customer/service/CustomerService.java
package com.hgw.gestionale.customer.service;

import com.hgw.gestionale.common.exception.NotFoundException;
import com.hgw.gestionale.customer.dto.CreateCustomerRequest;
import com.hgw.gestionale.customer.dto.CustomerResponse;
import com.hgw.gestionale.customer.dto.UpdateCustomerRequest;
import com.hgw.gestionale.customer.entity.Customer;
import com.hgw.gestionale.customer.mapper.CustomerMapper;
import com.hgw.gestionale.customer.repository.CustomerRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class CustomerService {
    private final CustomerRepository customerRepository;

    public CustomerResponse create(CreateCustomerRequest request) {
        Customer customer = CustomerMapper.toEntity(request);

        Customer savedCustomer = customerRepository.save(customer);

        return CustomerMapper.toResponse(savedCustomer);
    }

    public List<CustomerResponse> getAll() {
        return customerRepository.findAll()
                .stream()
                .map(CustomerMapper::toResponse)
                .toList();
    }

    public CustomerResponse getById(Long id) {
        Customer customer = customerRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Customer not found with id: " + id));

        return CustomerMapper.toResponse(customer);
    }

    public CustomerResponse update(Long id, UpdateCustomerRequest request) {
        Customer customer = customerRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Customer not found with id: " + id));

        CustomerMapper.updateEntity(customer, request);

        Customer updated = customerRepository.save(customer);

        return CustomerMapper.toResponse(updated);
    }

    public void delete(Long id) {
        Customer customer = customerRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Customer not found with id: " + id));

        customerRepository.deleteById(customer.getId());
    }

    public List<CustomerResponse> search(String nome, String cognome, String telefono, String email) {
        return customerRepository.search(nome, cognome, telefono, email)
                .stream()
                .map(CustomerMapper::toResponse)
                .toList();
    }
}

File: backend-node/src/service/customer.service.ts
import { and, eq, ilike, or } from "drizzle-orm";
import { db } from "../config/db";
import { CustomerResponse, CreateCustomerRequest, UpdateCustomerRequest } from "../dto/customer.dto";
import { CustomerMapper } from "../mapper/customer.mapper";
import { customers } from "../schema/customers";
import { HttpError } from "../common/httpError";
import { HttpStatus } from "../common/httpStatus";
import { SQL } from "drizzle-orm";

export class CustomerService {

    async create(request: CreateCustomerRequest): Promise<CustomerResponse> {
        const [saved] = await db.insert(customers).values({
            nome: request.nome,
            cognome: request.cognome,
            email: request.email,
            indirizzo: request.indirizzo,
            citta: request.citta,
            cap: request.cap,
            telefono: request.telefono,
            telefonoSecondario: request.telefonoSecondario,
        }).returning();

        return CustomerMapper.toResponse(saved);
    }

    async getAll(): Promise<CustomerResponse[]> {
        const result = await db.select().from(customers);
        return result.map(r => CustomerMapper.toResponse(r));
    }

    async search(nome?: string, cognome?: string, telefono?: string, email?: string): Promise<CustomerResponse[]> {
        const filters: SQL[] = [];
        if (nome) filters.push(ilike(customers.nome, `%${nome}%`));
        if (cognome) filters.push(ilike(customers.cognome, `%${cognome}%`));
        if (telefono) filters.push(ilike(customers.telefono, `%${telefono}%`));
        if (email) filters.push(ilike(customers.email, `%${email}%`));

        const result = await db.select().from(customers)
            .where(filters.length > 0 ? and(...filters) : undefined)
            .limit(10);

        return result.map(r => CustomerMapper.toResponse(r));
    }

    async getById(id: number): Promise<CustomerResponse> {
        const result = await db.select().from(customers).where(eq(customers.id, id));
        const customer = result.at(0);
        if (!customer) throw new HttpError(HttpStatus.NOT_FOUND, "Cliente non trovato");
        return CustomerMapper.toResponse(customer);
    }

    async update(id: number, request: UpdateCustomerRequest): Promise<CustomerResponse> {
        const result = await db.select().from(customers).where(eq(customers.id, id));
        if (!result.at(0)) throw new HttpError(HttpStatus.NOT_FOUND, "Cliente non trovato");

        const [updated] = await db.update(customers).set({
            nome: request.nome,
            cognome: request.cognome,
            email: request.email,
            indirizzo: request.indirizzo,
            citta: request.citta,
            cap: request.cap,
            telefono: request.telefono,
            telefonoSecondario: request.telefonoSecondario,
        }).where(eq(customers.id, id)).returning();

        return CustomerMapper.toResponse(updated);
    }

    async delete(id: number): Promise<void> {
        const result = await db.select().from(customers).where(eq(customers.id, id));
        if (!result.at(0)) throw new HttpError(HttpStatus.NOT_FOUND, "Cliente non trovato");

        await db.delete(customers).where(eq(customers.id, id));
    }

}
